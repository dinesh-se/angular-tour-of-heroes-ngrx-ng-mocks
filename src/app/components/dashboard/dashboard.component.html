<h2>Top Heroes</h2>

<div *ngIf="loading$ | async">Loading...</div>
<div *ngIf="(error$ | async) !== null">Error: {{ error$ | async }}</div>

<div class="heroes-menu">
  <a *ngFor="let hero of heroes$ | async | slice:1:5" routerLink="/detail/{{hero.id}}">
    {{hero.name}}
  </a>
</div>

<app-hero-search></app-hero-search>
