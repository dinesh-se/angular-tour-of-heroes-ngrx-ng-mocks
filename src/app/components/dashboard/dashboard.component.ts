import { AsyncPipe, NgFor, NgIf, SlicePipe } from '@angular/common';
import { Component, OnInit } from '@angular/core';
import { RouterLink } from '@angular/router';
import { HeroSearchComponent } from '@components/hero-search/hero-search.component';
import { Hero } from '@model/hero';
import { Store } from '@ngrx/store';
import { HeroActions } from '@store/hero.action';
import { HeroesState } from '@store/hero.reducer';
import { selectAllHeroes } from '@store/hero.selector';
import { Observable } from 'rxjs';

@Component({
  selector: 'app-dashboard',
  standalone: true,
  imports: [NgFor, RouterLink, HeroSearchComponent, AsyncPipe, NgIf, SlicePipe],
  templateUrl: './dashboard.component.html',
  styleUrl: './dashboard.component.scss'
})
export class DashboardComponent implements OnInit {
  heroes$!: Observable<Hero[]>;

  constructor(private store: Store<HeroesState>) {}

  ngOnInit(): void {
    this.store.dispatch(HeroActions.loadHeroes());
    this.heroes$ = this.store.select(selectAllHeroes);
  }
}
