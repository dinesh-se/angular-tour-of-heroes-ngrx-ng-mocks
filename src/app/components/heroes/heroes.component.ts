import { Component, ElementRef, OnInit, ViewChild } from '@angular/core';
import { AsyncPipe, NgFor, NgIf, UpperCasePipe } from '@angular/common';
import { RouterLink } from '@angular/router';

import { Hero } from '@model/hero';
import { HeroDetailComponent } from '@components/hero-detail/hero-detail.component';
import { Store } from '@ngrx/store';
import { HeroesState } from 'src/app/store/hero.reducer';
import { HeroActions } from '@store/hero.action';
import { Observable } from 'rxjs';
import { HeroSelectors } from '@store/hero.selector';

@Component({
  selector: 'app-heroes',
  standalone: true,
  imports: [
    UpperCasePipe,
    NgIf,
    NgFor,
    HeroDetailComponent,
    RouterLink,
    AsyncPipe
  ],
  templateUrl: './heroes.component.html',
  styleUrl: './heroes.component.scss'
})
export class HeroesComponent implements OnInit {
  heroes$!: Observable<Hero[]>;

  @ViewChild('heroName') heroName!: ElementRef;

  constructor(private store: Store<HeroesState>) {}

  ngOnInit(): void {
    this.store.dispatch(HeroActions.loadHeroes());
    this.heroes$ = this.store.select(HeroSelectors.selectAllHeroes);
  }

  add(name: string): void {
    name = name.trim();
    if (!name) { return; }
    this.store.dispatch(HeroActions.addHero({ name }));
    this.heroName.nativeElement.value = '';
  }

  delete(hero: Hero): void {
    this.store.dispatch(HeroActions.deleteHero({ id: hero.id }))
  }
}
