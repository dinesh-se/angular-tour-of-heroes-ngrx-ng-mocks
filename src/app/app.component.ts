import { Component } from '@angular/core';
import { RouterLink, RouterLinkActive, RouterOutlet } from '@angular/router';
import { HeroesComponent } from './components/heroes/heroes.component';
import { MessagesComponent } from '@components/messages/messages.component';

@Component({
  selector: 'app-root',
  standalone: true,
  imports: [
    RouterOutlet,
    HeroesComponent,
    MessagesComponent,
    RouterOutlet,
    RouterLink,
    RouterLinkActive
  ],
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss'
})
export class AppComponent {
  title = 'Tour of Heroes';
}
