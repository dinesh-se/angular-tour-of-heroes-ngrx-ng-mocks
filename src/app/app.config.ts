import { ApplicationConfig, importProvidersFrom, isDevMode } from '@angular/core';
import { provideRouter } from '@angular/router';
import { HttpClientInMemoryWebApiModule } from 'angular-in-memory-web-api';
import {provideHttpClient} from '@angular/common/http';
import { provideStoreDevtools } from '@ngrx/store-devtools';

import { InMemoryDataService } from '@services/in-memory-data.service';
import { routes } from './app.routes';
import { provideState, provideStore } from '@ngrx/store';
import { provideEffects } from '@ngrx/effects';
import { reducer } from '@store/hero.reducer';
import { HeroEffects } from '@store/hero.effects';

export const appConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes),
    provideHttpClient(),
    importProvidersFrom(HttpClientInMemoryWebApiModule.forRoot(InMemoryDataService, { dataEncapsulation: false })),
    provideStore(),
    provideState({ name: 'heroes', reducer }),
    provideEffects(HeroEffects),
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() })
]
};
