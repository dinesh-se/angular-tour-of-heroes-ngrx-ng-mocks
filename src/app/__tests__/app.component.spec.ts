import { fakeAsync, tick } from '@angular/core/testing';
import { Router, RouterModule, RouterOutlet } from '@angular/router';
import { Location } from '@angular/common';
import {
  MockBuilder,
  MockComponent,
  MockRender,
  NG_MOCKS_ROOT_PROVIDERS,
  ngMocks,
  MockedComponentFixture
} from 'ng-mocks';

import { DashboardComponent } from '@components/dashboard/dashboard.component';
import { HeroesComponent } from '@components/heroes/heroes.component';
import { AppComponent } from '../app.component';
import { routes } from '../app.routes';

describe('AppComponent Router', () => {
  beforeEach(async () => {
    await MockBuilder(
      [
        RouterModule.forRoot(routes),
        NG_MOCKS_ROOT_PROVIDERS,
        MockComponent(DashboardComponent),
        MockComponent(HeroesComponent)
      ],
    );
  });

  it('renders /dashboard with DashboardComponent', fakeAsync(() => {
    const fixture = MockRender(RouterOutlet, {});
    const router: Router = fixture.point.injector.get(Router);
    const location: Location = fixture.point.injector.get(Location);

    location.go('/dashboard');
    if (fixture.ngZone) {
      fixture.ngZone.run(() => router.initialNavigation());
      tick();
    }

    expect(location.path()).toEqual('/dashboard');
    expect(() => ngMocks.find(DashboardComponent)).not.toThrow();
  }));

  it('renders /heroes with HeroesComponent', fakeAsync(() => {
    const fixture = MockRender(RouterOutlet, {});
    const router: Router = fixture.point.injector.get(Router);
    const location: Location = fixture.point.injector.get(Location);

    location.go('/heroes');
    if (fixture.ngZone) {
      fixture.ngZone.run(() => router.initialNavigation());
      tick();
    }

    expect(location.path()).toEqual('/heroes');
    expect(() => ngMocks.find(HeroesComponent)).not.toThrow();
  }));
});

describe('AppComponent', () => {
  let fixture: MockedComponentFixture<AppComponent>;

  beforeEach(async () => {
    await MockBuilder(AppComponent);
    fixture = MockRender(AppComponent);
  });

  it('should create the app component', () => {
    const app = fixture.point.componentInstance;

    expect(app).toBeTruthy();
    expect(app.title).toEqual('Tour of Heroes');
    expect(fixture).toMatchSnapshot();
  });
});
