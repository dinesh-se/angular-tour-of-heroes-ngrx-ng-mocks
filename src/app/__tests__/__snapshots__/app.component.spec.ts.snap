// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should create the app component 1`] = `
<mock-render>
  <app-root>
    <h1>
      Tour of Heroes
    </h1>
    <nav>
      <a
        ariacurrentwhenactive="page"
        routerlink="/dashboard"
        routerlinkactive="active"
      >
        Dashboard
      </a>
      <a
        ariacurrentwhenactive="page"
        routerlink="/heroes"
        routerlinkactive="active"
      >
        Heroes
      </a>
    </nav>
    <router-outlet />
    <app-messages />
  </app-root>
</mock-render>
`;
